### PR TITLE
implement `From<Vec<Registers>>` to `ProcessorTable`

### PR DESCRIPTION
**Pull request by rkdud007**
resolving #68 

question 1: i'm not sure which step is handling adding & computing extension columns and also still not sure why we cannot have { }Table to contain full trace including registers + extension columns.

question 2: `ip` should change on padded rows  (base on Alan's tutorial) but our current machine return trace as keep last `ip` to be consistent on followup padded rows - is It expected? also added comment as //TODO for this

### CODE REVIEW

File: crates/brainfuck_prover/src/components/processor/table.rs [hunk 1]
```diff
@@ -78,13 +90,57 @@ impl ProcessorTable {
...
+    ///
+    /// Each dummy row increase clk, preserve ip, set others to zero.
+    ///
+    /// Does nothing if the table is empty.
+    fn pad(&mut self) {
+        if let Some(last_row) = self.table.last().cloned() {
+            let trace_len = self.table.len();
+            let padding_offset = (trace_len.next_power_of_two() - trace_len) as u32;
+            for i in 1..=padding_offset {
+                // TODO: i think ip should change in padding but trace returning like this
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/processor/table.rs:114]: Keeping the last `ip`is expected, as it is set as out of bound by the brainfuck_vm already.

The last step of a Brainfuck program execution trace provided by the `brainfuck_vm` crate is a row with `ci` and `ni` to 0 with `ip` out of bound (the length of the program)

For example, the program '++>,<[>+.<-]' executed on our `brainfuck_vm`, with the input 'a' provides the following trace (the same example is available in the [Neptune article](https://neptune.cash/learn/brainfuck-tutorial/)):
```
[
    clk:0, ip:0, ci:43, ni:43, mp:0, mv:0, mvi:0,
    clk:1, ip:1, ci:43, ni:62, mp:0, mv:1, mvi:1,
    clk:2, ip:2, ci:62, ni:44, mp:0, mv:2, mvi:1073741824,
    clk:3, ip:3, ci:44, ni:60, mp:1, mv:0, mvi:0,
    clk:4, ip:4, ci:60, ni:91, mp:1, mv:97, mvi:2081066627,
    clk:5, ip:5, ci:91, ni:13, mp:0, mv:2, mvi:1073741824,
    clk:6, ip:7, ci:62, ni:43, mp:0, mv:2, mvi:1073741824,
    clk:7, ip:8, ci:43, ni:46, mp:1, mv:97, mvi:2081066627,
    clk:8, ip:9, ci:46, ni:60, mp:1, mv:98, mvi:898437036,
    clk:9, ip:10, ci:60, ni:45, mp:1, mv:98, mvi:898437036,
    clk:10, ip:11, ci:45, ni:93, mp:0, mv:2, mvi:1073741824,
    clk:11, ip:12, ci:93, ni:7, mp:0, mv:1, mvi:1,
    clk:12, ip:7, ci:62, ni:43, mp:0, mv:1, mvi:1,
    clk:13, ip:8, ci:43, ni:46, mp:1, mv:98, mvi:898437036,
    clk:14, ip:9, ci:46, ni:60, mp:1, mv:99, mvi:2125791893,
    clk:15, ip:10, ci:60, ni:45, mp:1, mv:99, mvi:2125791893,
    clk:16, ip:11, ci:45, ni:93, mp:0, mv:1, mvi:1,
    clk:17, ip:12, ci:93, ni:7, mp:0, mv:0, mvi:0,
    clk:18, ip:14, ci:0, ni:0, mp:0, mv:0, mvi:0,
]
```
The padded rows should simply  increments the `clk` (representings the steps):
It's actually what is done in the python implementation of Alan ([here](https://github.com/aszepieniec/stark-brainfuck/blob/4c955cf34d2eed0a862e8ffeba4ebfa3b24d1ec9/code/processor_table.py#L24-L35))

```
    clk:19, ip:14, ci:0, ni:0, mp:0, mv:0, mvi:0,
    clk:20, ip:14, ci:0, ni:0, mp:0, mv:0, mvi:0,
    ...
```

File: crates/brainfuck_prover/src/components/processor/table.rs [hunk 2]
```diff
@@ -34,6 +34,18 @@ pub struct ProcessorTableRow {
     mvi: BaseField,
 }
 
+impl ProcessorTableRow {
+    /// Creates a row for the [`MemoryTable`] which is considered 'dummy'.
+    ///
+    /// A 'dummy' row, is a row that is not part of the execution trace from the Brainfuck program
+    /// execution.
+    /// They are used for padding and filling the `clk` gaps after sorting by `mp`, to enforce the
+    /// correct sorting.
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/processor/table.rs:43]: The processor table is only sorted by `clk`

File: crates/brainfuck_prover/src/components/processor/table.rs [hunk 3]
```diff
@@ -78,13 +90,57 @@ impl ProcessorTable {
...
+    /// * `rows` - A vector of [`ProcessorTableRow`] to add to the table.
+    ///
+    /// This method extends the `table` vector with the provided rows.
+    fn add_rows(&mut self, rows: Vec<ProcessorTableRow>) {
+        self.table.extend(rows);
+    }
+
+    /// Pads the Processor table with dummy rows up to the next power of two length.
+    ///
+    /// Each dummy row increase clk, preserve ip, set others to zero.
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/processor/table.rs:106]: Actually it is 'increase clk, copy the others from the last step (i.e. last row of the table)

File: crates/brainfuck_prover/src/components/processor/table.rs [hunk 4]
```diff
@@ -78,13 +90,57 @@ impl ProcessorTable {
...
+    ///
+    /// Does nothing if the table is empty.
+    fn pad(&mut self) {
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/processor/table.rs:109]: For this function, the logic is done [here](https://github.com/kkrt-labs/stwo-brainfuck/blob/9e5eb395e1787b48a6d6882903b9c5816bf8a93d/crates/brainfuck_vm/src/machine.rs#L244-L255) in the brainfuck_vm

You can refactor it to the ProcessorTable (we're actually not padding the trace in the brainfuck_vm crate)

**Review by zmalatrax — COMMENTED** [review]
Nice, answered question 2 in a review comment :)

File: crates/brainfuck_prover/src/components/processor/table.rs (hunk 1 above)
Comment (rkdud007) [reply crates/brainfuck_prover/src/components/processor/table.rs:114]: oh i see makesense thought clk 18 is padding

File: crates/brainfuck_prover/src/components/processor/table.rs (hunk 4 above)
Comment (rkdud007) [reply crates/brainfuck_prover/src/components/processor/table.rs:109]: so this https://github.com/kkrt-labs/stwo-brainfuck/blob/9435e6c50cc935959cbd14c2a9a2bce0ca3520c8/crates/brainfuck_vm/src/bin/brainfuck_vm.rs#L49 part can be ok to be removed?

Comment (zmalatrax) [reply crates/brainfuck_prover/src/components/processor/table.rs:109]: yeah, along with the pad_trace flag

File: crates/brainfuck_vm/src/bin/brainfuck_vm.rs [hunk 5]
```diff
@@ -45,9 +45,6 @@ fn main() -> Result<(), MachineError> {
     tracing::info!("Provide inputs separated by linefeeds: ");
     bf_vm.execute().unwrap();
     if args.trace {
-        if args.pad_trace {
```
Comment (zmalatrax) [inline crates/brainfuck_vm/src/bin/brainfuck_vm.rs:48]: Also remove the pad_trace argument in the `Args` struct (cannot comment on it for the review): L23-L34 of this file

**Comment by rkdud007** [comment]
i did `trunk check` and `trunk fmt --all`, is there smth I missed in terms of first CI?

File: crates/brainfuck_prover/src/components/processor/table.rs [hunk 6]
```diff
@@ -78,13 +90,56 @@ impl ProcessorTable {
...
+        if let Some(last_row) = self.table.last().cloned() {
+            let trace_len = self.table.len();
+            let padding_offset = (trace_len.next_power_of_two() - trace_len) as u32;
+            for i in 1..=padding_offset {
+                self.add_row(ProcessorTableRow::new_dummy(
+                    last_row.clk + BaseField::from(i),
+                    last_row.ip,
+                ));
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/processor/table.rs:117]: The padding logic is different from the header doc (and the `pad_trace` deleted in the `brainfuck_vm` crate)
As we're only increasing the `clk` register and copying the last instructions' other registers, it's not required to have a `new_dummy` function for the padding imo

**Comment by zmalatrax** [comment]
> i did `trunk check` and `trunk fmt --all`, is there smth I missed in terms of first CI?

It seems to be clippy related, does trunk effectively runs clippy?
(You can also run clippy directly `cargo +nightly-2024-01-04 clippy` (we're stuck with this nightly version as Stwo has not merged their toolchain update)

Maybe you have a notice similar to this:
> ~/kkrt/stwo-brainfuck main *5 ❯ trunk check 
Checking 100% [=================================================================>]  25/25  1.1s 
>                                                                                                                                                   
>  NOTICES  
> 'rust' has no matching download for current os/cpu @ version:2024-01-04

File: crates/brainfuck_prover/src/components/processor/table.rs [hunk 7]
```diff
@@ -78,13 +78,57 @@ impl ProcessorTable {
...
+    /// Does nothing if the table is empty.
+    fn pad(&mut self) {
+        if let Some(last_row) = self.table.last().cloned() {
+            let trace_len = self.table.len();
+            let padding_offset = (trace_len.next_power_of_two() - trace_len) as u32;
+            for i in 1..=padding_offset {
+                self.add_row(ProcessorTableRow {
+                    clk: last_row.clk + BaseField::from(i),
+                    ip: last_row.ip,
+                    ..Default::default()
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/processor/table.rs:105]: Even though we expect `ci` and `ni` to be the default values (0) for the last row showing the end of the program, the `mv` and `mvi` fields might not be the default ones (if the last memory address pointed by `mp` has been incremented)

```suggestion
                    ..last_row.clone()
```

**Review by zmalatrax — APPROVED** [review]

**Pull request merged**